File: src/com/designing/singleton/java/SingletonSafe_LH.java
package com.designing.singleton.java;

/**
 * 懒汉模式  加锁保证线程安全
 */
public class SingletonSafe_LH {
    private static SingletonSafe_LH instance;
    // 使用同步锁
    public static synchronized SingletonSafe_LH getInstance(){
        if (instance == null){
            instance = new SingletonSafe_LH();
        }
        return instance;
    }
}

File: src/com/designing/singleton/kotlin/SingletonSafe_LH.kt
package com.designing.singleton.kotlin

// 懒汉 (线程安全)
class SingletonSafe_LH private constructor(){
    companion object {
        private var instance : SingletonSafe_LH? = null
        get() {
            if (field == null){
                field = SingletonSafe_LH()
            }
            return field
        }
        @Synchronized
        fun get() : SingletonSafe_LH{
            return instance!!
        }
    }
}